### PR TITLE
Reduce Allocation Merges Updates

### PR DESCRIPTION
**Pull request by JohnTortugo**
Please, review this PR with the backport of the following fixes:

- https://github.com/openjdk/jdk/pull/19545
- https://github.com/openjdk/jdk/pull/19148
- https://github.com/openjdk/jdk/pull/18851
- https://github.com/openjdk/jdk/pull/20087

Tested locally on Linux x86_64.

### CODE REVIEW

File: src/hotspot/share/opto/macro.cpp [hunk 1]
```diff
@@ -651,6 +651,9 @@ bool PhaseMacroExpand::can_eliminate_allocation(PhaseIterGVN* igvn, AllocateNode
     if (res_type == NULL) {
       NOT_PRODUCT(fail_eliminate = "Neither instance or array allocation";)
       can_eliminate = false;
+    } else if (!res_type->klass_is_exact()) {
```
Comment (brianjstafford) [inline src/hotspot/share/opto/macro.cpp:654]: Where is this code in the tip PRs?

Comment (JohnTortugo) [reply src/hotspot/share/opto/macro.cpp:654]: https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/macro.cpp#L576

Comment (JohnTortugo) [reply src/hotspot/share/opto/macro.cpp:654]: This is the PR that added the change: https://github.com/openjdk/jdk/pull/18851

**Comment by brianjstafford** [comment]
I don't see the updated AllocationMergesTests.java file in any of the PRs.

**Comment by JohnTortugo** [comment]
> I don't see the updated AllocationMergesTests.java file in any of the PRs.

JDK11 doesn't support the IR-based testing framework. That's why that file doesn't exist in this branch/repro.

File: src/hotspot/share/opto/macro.cpp (hunk 1 above)
Comment (brianjstafford) [reply src/hotspot/share/opto/macro.cpp:654]: Ah - can you please add that to the list of PRs you've backported then?

Comment (brianjstafford) [reply src/hotspot/share/opto/macro.cpp:654]: You'll need to update the descriptions in the other PRs as well.

Comment (JohnTortugo) [reply src/hotspot/share/opto/macro.cpp:654]: I did that, perhaps at the same time that you were typing here :-)

**Review by brianjstafford — APPROVED** [review]
LGTM
Make sure to update the patch index file afterwards as well.

**Pull request merged**